contracts/token/BiconomyTokenPaymaster.sol

```diff
@@ -0,0 +1,563 @@
+import { SafeTransferLib } from "@solady/src/utils/SafeTransferLib.sol";
+import { BasePaymaster } from "../base/BasePaymaster.sol";
+import { BiconomyTokenPaymasterErrors } from "../common/BiconomyTokenPaymasterErrors.sol";
+import { IBiconomyTokenPaymaster } from "../interfaces/IBiconomyTokenPaymaster.sol";
+import { IOracle } from "../interfaces/oracles/IOracle.sol";
+import { TokenPaymasterParserLib } from "../libraries/TokenPaymasterParserLib.sol";
+import { SignatureCheckerLib } from "@solady/src/utils/SignatureCheckerLib.sol";
+import { ECDSA as ECDSA_solady } from "@solady/src/utils/ECDSA.sol";
+import "@account-abstraction/contracts/core/Helpers.sol";
+import "./swaps/Uniswapper.sol";
+
+/**
+ * @title BiconomyTokenPaymaster
+ * @author ShivaanshK<shivaansh.kapoor@biconomy.io>
+ * @author livingrockrises<chirag@biconomy.io>
+ * @notice Biconomy's Token Paymaster for Entry Point v0.7
+ * @dev  A paymaster that allows user to pay gas fees in ERC20 tokens. The paymaster uses the precharge and refund model
+ * to handle gas remittances.
+ *
+ * Currently, the paymaster supports two modes:
+ * 1. EXTERNAL - Relies on a quoted token price from a trusted entity (verifyingSigner).
+ * 2. INDEPENDENT - Relies purely on price oracles (Offchain and TWAP) which implement the IOracle interface. This mode
+ * doesn't require a signature and is always "available" to use.
+ *
+ * The paymaster's owner has full discretion over the supported tokens (for independent mode), price adjustments
+ * applied, and how
+ * to manage the assets received by the paymaster.
+ */
+contract BiconomyTokenPaymaster is
+    IBiconomyTokenPaymaster,
+    BasePaymaster,
+    ReentrancyGuardTransient,
+    BiconomyTokenPaymasterErrors,
+    Uniswapper
+{
+    using UserOperationLib for PackedUserOperation;
+    using TokenPaymasterParserLib for bytes;
+    using SignatureCheckerLib for address;
+
+    // State variables
+    address public verifyingSigner;
+    uint256 public unaccountedGas;
+    uint256 public priceMarkup;
+    uint256 public priceExpiryDuration;
+    IOracle public nativeAssetToUsdOracle; // ETH -> USD price oracle
+    mapping(address => TokenInfo) tokenDirectory;
+
+    // PAYMASTER_ID_OFFSET
+    uint256 private constant UNACCOUNTED_GAS_LIMIT = 50_000; // Limit for unaccounted gas cost
+    uint256 private constant PRICE_DENOMINATOR = 1e6; // Denominator used when calculating cost with price markup
+    uint256 private constant MAX_PRICE_MARKUP = 2e6; // 100% premium on price (2e6/PRICE_DENOMINATOR)
+
+    constructor(
+        address _owner,
+        address _verifyingSigner,
+        IEntryPoint _entryPoint,
+        uint256 _unaccountedGas,
+        uint256 _priceMarkup,
+        uint256 _priceExpiryDuration,
+        IOracle _nativeAssetToUsdOracle,
+        ISwapRouter _uniswapRouter,
+        address _wrappedNative,
+        address[] memory _tokens, // Array of token addresses supported by the paymaster
```

Comment:
these are like the bucket of tokens which are supported in independent mode. think of corresponding name or maybe just comment.

Reply:
agreed. Needs a better name.